File: Api/Services/AI/OpenAiProvider.cs
using Microsoft.Extensions.Options;
using OpenAI.Chat;
using LegalStrategyAdvisor.Api.Configuration;
using LegalStrategyAdvisor.Api.Services.AI;

namespace LegalStrategyAdvisor.Api.Services.AI;

public class OpenAiProvider : IAiProvider
{
    private readonly OpenAIOptions _options;
    private readonly ChatClient _chatClient;
    private readonly ILogger<OpenAiProvider> _logger;

    public string ProviderName => "OpenAI";

    public OpenAiProvider(IOptions<AiProviderOptions> options, ILogger<OpenAiProvider> logger)
    {
        _options = options.Value.OpenAI;
        _logger = logger;

        if (string.IsNullOrEmpty(_options.ApiKey))
        {
            throw new InvalidOperationException("OpenAI API key is not configured");
        }

        try
        {
            var client = new OpenAI.OpenAIClient(_options.ApiKey);
            _chatClient = client.GetChatClient(_options.Model);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to initialize OpenAI client");
            throw new AiException(ProviderName, "Failed to initialize OpenAI client", ex);
        }
    }

    public async Task<string> GenerateResponseAsync(string prompt, CancellationToken cancellationToken = default)
    {
        try
        {
            _logger.LogDebug("Sending request to OpenAI with model: {Model}", _options.Model);

            var messages = new List<ChatMessage>
            {
                new SystemChatMessage("You are a legal strategy advisor AI assistant. Provide comprehensive, well-structured legal analysis and strategy recommendations. Always include relevant legal precedents, potential risks, and actionable steps. Format your response clearly with sections and bullet points where appropriate."),
                new UserChatMessage(prompt)
            };

            var chatCompletionOptions = new ChatCompletionOptions
            {
                MaxOutputTokenCount = _options.MaxTokens,
                Temperature = (float)_options.Temperature
            };

            var response = await _chatClient.CompleteChatAsync(messages, chatCompletionOptions, cancellationToken);

            if (response?.Value?.Content?.Count > 0)
            {
                var content = string.Join("\n", response.Value.Content.Select(c => c.Text));

                _logger.LogInformation("OpenAI request completed successfully. Tokens used: {TokensUsed}",
                    response.Value.Usage?.TotalTokenCount ?? 0);

                return content;
            }

            throw new AiException(ProviderName, "Empty response received from OpenAI");
        }
        catch (OperationCanceledException)
        {
            _logger.LogWarning("OpenAI request was cancelled");
            throw;
        }
        catch (Exception ex) when (ex is not AiException)
        {
            _logger.LogError(ex, "Error generating response from OpenAI");
            throw new AiException(ProviderName, $"OpenAI request failed: {ex.Message}", ex);
        }
    }

    public async Task<bool> IsAvailableAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            // Test with a minimal request
            var testMessage = new UserChatMessage("Hello");
            var testOptions = new ChatCompletionOptions { MaxOutputTokenCount = 10 };

            var response = await _chatClient.CompleteChatAsync([testMessage], testOptions, cancellationToken);
            return response?.Value != null;
        }
        catch (Exception ex)
        {
            _logger.LogWarning(ex, "OpenAI availability check failed");
            return false;
        }
    }
}

File: AiService/providers/openai_provider.py
import os
import openai
from pydantic import BaseModel
from typing import Optional
import logging

logger = logging.getLogger(__name__)

class OpenAIConfig(BaseModel):
    api_key: str
    model: str = "gpt-4o-mini"
    max_tokens: int = 512
    temperature: float = 0.7

class OpenAIProvider:
    def __init__(self):
        self.config = OpenAIConfig(
            api_key=os.environ.get("OPENAI_API_KEY", ""),
            model=os.environ.get("OPENAI_MODEL", "gpt-4o-mini"),
            max_tokens=int(os.environ.get("OPENAI_MAX_TOKENS", "512")),
            temperature=float(os.environ.get("OPENAI_TEMPERATURE", "0.7"))
        )
        
        if self.config.api_key:
            openai.api_key = self.config.api_key
            logger.info(f"OpenAI provider initialized with model: {self.config.model}")
        else:
            logger.warning("OpenAI API key not provided")

    async def generate(self, prompt: str) -> str:
        """Generate AI response using OpenAI API"""
        if not self.config.api_key:
            return "[Error: OpenAI API key missing]"
        
        try:
            # Use the newer client pattern
            client = openai.AsyncOpenAI(api_key=self.config.api_key)
            
            system_prompt = """You are a legal strategy advisor AI assistant. 
            Provide comprehensive, well-structured legal analysis and strategy recommendations. 
            Always include relevant legal precedents, potential risks, and actionable steps. 
            Format your response clearly with sections and bullet points where appropriate.
            Keep responses professional and within the specified token limits."""
            
            response = await client.chat.completions.create(
                model=self.config.model,
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": prompt}
                ],
                max_tokens=self.config.max_tokens,
                temperature=self.config.temperature,
                timeout=30.0  # 30 second timeout
            )
            
            content = response.choices[0].message.content
            if content:
                logger.info(f"OpenAI response generated successfully. Tokens used: {response.usage.total_tokens if response.usage else 'unknown'}")
                return content.strip()
            else:
                return "[Error: Empty response from OpenAI]"
                
        except openai.RateLimitError as e:
            logger.error(f"OpenAI rate limit error: {e}")
            return "[Error: Rate limit exceeded. Please try again later.]"
        except openai.AuthenticationError as e:
            logger.error(f"OpenAI authentication error: {e}")
            return "[Error: Invalid API key]"
        except openai.APITimeoutError as e:
            logger.error(f"OpenAI timeout error: {e}")
            return "[Error: Request timeout]"
        except Exception as e:
            logger.error(f"OpenAI error: {e}")
            return f"[OpenAI error: {str(e)}]"

    def is_available(self) -> bool:
        """Check if OpenAI provider is properly configured"""
        return bool(self.config.api_key)
